backend/src/controllers/v2/environmentController.ts

```diff
@@ -0,0 +1,167 @@
+import { Request, Response } from 'express';
+import * as Sentry from '@sentry/node';
+import { Secret, ServiceToken, Workspace, Integration } from '../../models';
+
+/**
+ * Create new workspace environment named [environmentName] under workspace with id
+ * @param req
+ * @param res
+ * @returns
+ */
+export const createWorkspaceEnvironment = async (
+  req: Request,
+  res: Response
+) => {
+  const { workspaceId, environmentName, environmentSlug } = req.body;
+  try {
+    // atomic create the environment
+    const workspace = await Workspace.findOneAndUpdate(
+      {
+        _id: workspaceId,
+        'environments.slug': { $ne: environmentSlug },
+        'environments.name': { $ne: environmentName },
+      },
+      {
+        $addToSet: {
+          environments: { name: environmentName, slug: environmentSlug },
+        },
+      }
+    );
+
+    if (!workspace) {
+      throw new Error('Failed to update workspace environment');
+    }
+  } catch (err) {
+    Sentry.setUser({ email: req.user.email });
+    Sentry.captureException(err);
+    return res.status(400).send({
+      message: 'Failed to create new workspace environment',
+    });
+  }
+
+  return res.status(200).send({
+    message: 'Successfully created new environment',
+    workspace: workspaceId,
+    environment: {
+      name: environmentName,
+      slug: environmentSlug,
+    },
+  });
+};
+
+/**
+ * Rename workspace environment with new name and slug of a workspace with [workspaceId]
+ * Old slug [oldEnvironmentSlug] must be provided
+ * @param req
+ * @param res
+ * @returns
+ */
+export const renameWorkspaceEnvironment = async (
+  req: Request,
+  res: Response
+) => {
+  const { workspaceId, environmentName, environmentSlug, oldEnvironmentSlug } =
+    req.body;
+  try {
+    // user should pass both new slug and env name
+    if (!environmentSlug || !environmentName) {
+      throw new Error('Invalid environment given.');
+    }
+
+    // atomic update the env to avoid conflict
+    const workspace = await Workspace.findOneAndUpdate(
+      { _id: workspaceId, 'environments.slug': oldEnvironmentSlug },
+      {
+        'environments.$.name': environmentName,
+        'environments.$.slug': environmentSlug,
+      }
+    );
+    if (!workspace) {
+      throw new Error('Failed to update workspace');
+    }
+
+    await Secret.updateMany(
+      { workspace: workspaceId, environment: oldEnvironmentSlug },
+      { environment: environmentSlug }
+    );
+    await ServiceToken.updateMany(
```

Comment:
Above and following database operations should follow `Atomicity` from ACID. This applies for similar operations further in the PR.

Reply:
Can you be more elaborate your point. All the above operations are atomic by nature from Mongodb 4.0 onwards.

Reply:
@Zamion101 Ohhh sure that's valid issue. But the catch is for mongodb to support transactions, its not simple as sql db. U need to enable replication to support transactions. For that we need to change the image's doc and also prod i guess.

I think then the scope gets bigger for this PR. Would love to get input from @dangtony98 @maidul98 on this.

Reply:
I don't think we will have any problem as it is right now. I checked the dependencies and package versions and we're good to use following method from Mongoose documents;

https://mongoosejs.com/docs/transactions.html

Reply:
@Zamion101 Yah library wise is ok all good.

But to support transactions in mongodb there is a architecture change in db required.

Check this: https://www.mongodb.com/community/forums/t/why-replica-set-is-mandatory-for-transactions-in-mongodb/9533